### ISSUE TITLE
Do not call `write` with empty strings

### ISSUE THREAD

**Issue by cipriancraciun** [djc/askama#318]
While debugging the efficient sub-template rendering I've stumbled upon the following strange behaviour:  the `write` method of `fmt::Write` is called multiple times with an empty string.

Granted this has a minor performance impact especially if buffered writers are used;  however the behaviour is strange.

Bellow the following is the template:
~~~~
{% for index in 0..100 -%}
	{{- "abc" -}}
{%- endfor %}
~~~~

And the resulting code:
~~~~
impl ::askama::Template for TestTemplate {
    fn render_into(&self, writer: &mut ::std::fmt::Write) -> ::askama::Result<()> {
        include_bytes !
("/tmp/test.html")
;
        for (index, _loop_item) in ::askama::helpers::TemplateLoop::new(0..100) {
            write!(
                writer,
                "{expr0}",
                expr0 = &::askama::MarkupDisplay::new_unsafe(&"abc", ::askama::Html),
            )?;
        }
        Ok(())
    }
    fn extension(&self) -> Option<&'static str> {
        Some("html")
    }
    fn size_hint(&self) -> usize {
        0
    }
}
impl ::askama::SizedTemplate for TestTemplate {
    fn size_hint() -> usize {
        0
    }
    fn extension() -> Option<&'static str> {
        Some("html")
    }
}
impl ::std::fmt::Display for TestTemplate {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        ::askama::Template::render_into(self, f).map_err(|_| ::std::fmt::Error {})
    }
}
~~~~

And bellow is the call trace to a custom `fmt::Write`:
~~~~
struct Writer {}

impl fmt::Write for Writer {
    fn write_str (&mut self, _string : &str) -> (Result<(), fmt::Error>) {
        eprintln! ("[>>] {} | {}", _string.len (), _string);
        return Ok (());
    }
}
~~~~
~~~~
[>>] 0 |
[>>] 0 |
[>>] 3 | abc
[>>] 0 | 
[>>] 0 | 
[>>] 3 | abc
[>>] 0 | 
[>>] 0 | 
[>>] 3 | abc
[>>] 0 | 
[>>] 0 | 
[>>] 3 | abc
...
~~~~

As can be seen, there are many calls with an empty string.

**Comment by djc** [comment]
Well, from the generated code it seems to me like there is no code generating empty writes, so I'm guessing this happens at an implementation layer below what Askama is targeting.

**Comment by cipriancraciun** [comment]
> [...]  so I'm guessing this happens at an implementation layer below what Askama is targeting.

For some reason this is a direct consequence of Askama, because if I add or remove `{- -}` markers, it generates or not the empty strings.

Also the custom writer I'm using does nothing fancy except logging each individual `write` call.

So given that the above call uses nothing else except `String` (that should behave identically with regard to `Display`), and the own Askama code, the extra writes come must from within Askama's own code.

**Comment by djc** [comment]
Huh, that's weird then. If you can investigate, that would be great.

**Comment by cipriancraciun** [comment]
OK, I've made a minor change to the `io::Write` adapter so I can quantify how many empty strings we are actually pushing to the sink:
* for the `Big table` apparently 24%;
* for the `Testing teams` also 24%;
* for the tests somewhere around 35%;

In conclusion, for every 3 "useful" writes, we make another one with an empty string.

(It would be helpful to include in the tests also some real-world-like templates, because these are perhaps corner-cases.)

----

I've tried to track down the issue, and I think it is embedded deep into the `EscapeWriter`.

After applying the following patch, the number of empty writes has went down to 0 for the benchmarks.

In terms of benchmarks the impact is quite big:
* for the `Big table` around 10%;
* for the `Teams` around 15%;
* (all these figures are on-top of the "generic" write implementation that doesn't use `&mut dyn Write` instances;  if we use the plain `dyn` implementation, then the figures are much higher;)

~~~~diff
diff --git i/askama_escape/src/lib.rs w/askama_escape/src/lib.rs
index 2771507..bda153f 100644
--- i/askama_escape/src/lib.rs
+++ w/askama_escape/src/lib.rs
@@ -122,32 +122,40 @@ impl Escaper for Html {
                     b'<' => escaping_body!(start, i, fmt, bytes, "&lt;"),
                     b'>' => escaping_body!(start, i, fmt, bytes, "&gt;"),
                     b'&' => escaping_body!(start, i, fmt, bytes, "&amp;"),
                     b'"' => escaping_body!(start, i, fmt, bytes, "&quot;"),
                     b'\'' => escaping_body!(start, i, fmt, bytes, "&#x27;"),
                     b'/' => escaping_body!(start, i, fmt, bytes, "&#x2f;"),
                     _ => (),
                 }
             }
         }
-        fmt.write_str(unsafe { str::from_utf8_unchecked(&bytes[start..]) })
+        if start < bytes.len() {
+            fmt.write_str(unsafe { str::from_utf8_unchecked(&bytes[start..]) })
+        } else {
+            Ok(())
+        }
     }
 }
 
 pub struct Text;
 
 impl Escaper for Text {
     fn write_escaped<W>(&self, mut fmt: W, string: &str) -> fmt::Result
     where
         W: Write,
     {
-        fmt.write_str(string)
+        if !string.is_empty() {
+            fmt.write_str(string)
+        } else {
+            Ok(())
+        }
     }
 }
 
 #[derive(Debug, PartialEq)]
 enum DisplayValue<T>
 where
     T: Display,
 {
     Safe(T),
     Unsafe(T),
~~~~

----

@djc Should I submit a pull request with these changes?

**Comment by djc** [comment]
Yes, a PR would be great! Does the `Text` variant actually have an impact? That part seems like it would be glossing over an underlying problem.

**Comment by cipriancraciun** [comment]
> Does the `Text` variant actually have an impact?

Given that we don't have any meaningful text only templates, I would say it's unknown, however I would include it for symmetry with the HTML one, and as a reminder for implementers to skip on empty strings.

**Comment by djc** [comment]
I would say it's incorrect. If you have found that the escaping code is what introduces the empty writes, then it follows that non-escaping code doesn't have any. Let's skip the `Text` part, then.

**Comment by cipriancraciun** [comment]
OK, I've removed it from the text escaper.

However I would propose amending my `io::Write` implementation in the other pull request to always ignore empty writes.  (Given that those functions might be used with undecorated files or network streams, empty writes are never a good idea...)

**Comment by djc** [comment]
Sure, that sounds good.

**Comment by cipriancraciun** [comment]
It was solved by PR #325. Closing.
